tests/odb/freshen.c

```diff
@@ -17,34 +17,99 @@ void test_odb_freshen__cleanup(void)
+
+	old[0].tv_sec = 1234567890;
+	old[0].tv_usec = 0;
+	old[1].tv_sec = 1234567890;
+	old[1].tv_usec = 0;
+
+	git_buf_joinpath(&fullpath, "testrepo.git/objects", fn);
+
+	cl_must_pass(p_utimes(git_buf_cstr(&fullpath), old));
+	cl_must_pass(p_lstat(git_buf_cstr(&fullpath), out));
+	git_buf_free(&fullpath);
+}
 
-void test_odb_freshen__loose_object(void)
+#define LOOSE_STR     "my new file\n"
+#define LOOSE_BLOB_ID "a71586c1dfe8a71c6cbf6c129f404c5642ff31bd"
+#define LOOSE_BLOB_FN "a7/1586c1dfe8a71c6cbf6c129f404c5642ff31bd"
+
+void test_odb_freshen__loose_blob(void)
 {
 	git_oid expected_id, id;
 	struct stat before, after;
-	struct p_timeval old_times[2];
 
-	cl_git_pass(git_oid_fromstr(&expected_id, LOOSE_ID));
+	cl_git_pass(git_oid_fromstr(&expected_id, LOOSE_BLOB_ID));
+	set_time_wayback(&before, LOOSE_BLOB_FN);
 
-	old_times[0].tv_sec = 1234567890;
-	old_times[0].tv_usec = 0;
-	old_times[1].tv_sec = 1234567890;
-	old_times[1].tv_usec = 0;
+	/* make sure we freshen a blob */
+	cl_git_pass(git_blob_create_frombuffer(&id, repo, LOOSE_STR, CONST_STRLEN(LOOSE_STR)));
+	cl_assert_equal_oid(&expected_id, &id);
+	cl_must_pass(p_lstat("testrepo.git/objects/" LOOSE_BLOB_FN, &after));
 
-	/* set time to way back */
-	cl_must_pass(p_utimes("testrepo.git/objects/" LOOSE_FN, old_times));
-	cl_must_pass(p_lstat("testrepo.git/objects/" LOOSE_FN, &before));
+	cl_assert(before.st_atime < after.st_atime);
+	cl_assert(before.st_mtime < after.st_mtime);
+}
+
+#define LOOSE_TREE_ID "944c0f6e4dfa41595e6eb3ceecdb14f50fe18162"
+#define LOOSE_TREE_FN "94/4c0f6e4dfa41595e6eb3ceecdb14f50fe18162"
+
+void test_odb_freshen__loose_tree(void)
+{
+	git_oid expected_id, id;
+	git_tree *tree;
+	struct stat before, after;
+
+	cl_git_pass(git_oid_fromstr(&expected_id, LOOSE_TREE_ID));
+	set_time_wayback(&before, LOOSE_TREE_FN);
+
+	cl_git_pass(git_tree_lookup(&tree, repo, &expected_id));
+	cl_git_pass(git_tree_create_updated(&id, repo, tree, 0, NULL));
 
-	cl_git_pass(git_odb_write(&id, odb, LOOSE_STR, CONST_STRLEN(LOOSE_STR),
-		GIT_OBJ_BLOB));
+	/* make sure we freshen a tree */
 	cl_assert_equal_oid(&expected_id, &id);
-	cl_must_pass(p_lstat("testrepo.git/objects/" LOOSE_FN, &after));
+	cl_must_pass(p_lstat("testrepo.git/objects/" LOOSE_TREE_FN, &after));
+
+	cl_assert(before.st_atime < after.st_atime);
+	cl_assert(before.st_mtime < after.st_mtime);
+
+	git_tree_free(tree);
+}
+
+void test_odb_freshen__tree_during_commit(void)
+{
+	git_oid tree_id, parent_id, commit_id;
+	git_tree *tree;
+	git_commit *parent;
+	git_signature *signature;
+	struct stat before, after;
+
+	cl_git_pass(git_oid_fromstr(&tree_id, LOOSE_TREE_ID));
+	cl_git_pass(git_tree_lookup(&tree, repo, &tree_id));
+	set_time_wayback(&before, LOOSE_TREE_FN);
 
+	cl_git_pass(git_oid_fromstr(&parent_id, "a65fedf39aefe402d3bb6e24df4d4f5fe4547750"));
+	cl_git_pass(git_commit_lookup(&parent, repo, &parent_id));
+
+	cl_git_pass(git_signature_new(&signature,
+		"Refresher", "refresher@refresh-me.com", 1488547083, 0));
```

Comment:
Haha.  Right, yes, I probably shouldn't use stream of consciousness for things that are etched in git's commit history.